x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java

```diff
@@ -310,62 +294,89 @@ public Integer getScrollSize() {
         return scrollSize;
     }
 
-    public QueryBuilder getParsedQuery() {
-        return querySupplier.get();
+    public QueryBuilder getParsedQuery(NamedXContentRegistry namedXContentRegistry) {
+        return parseQuery(namedXContentRegistry, new ArrayList<>());
+    }
+
+    private QueryBuilder parseQuery(NamedXContentRegistry namedXContentRegistry, List<String> deprecations) {
+        try {
+            return XContentObjectTransformer.queryBuilderTransformer(namedXContentRegistry).fromMap(queryProvider.query, deprecations);
```

Comment:
`queryProvider` has the parsed query (`QueryBuilder`), is it necessary to parse it again?

(depredations are ignored on line 314 anyway)

Reply:
Yes, the only reason to have the parsed one kept is when serializing to a node that does not support lazily parsed objects.

Reply:
does that mean, `QueryBuilder.parsedQuery` could be null while query contains a __valid__ query?

Reply:
@hendrikmuhs This would only be the case if there were 3 versions of Elastic search running in the cluster, which is unsupported. I am going to put an early failure in serialization if Parsed == null && Exception == null when shipping to an "Eager" node. 

Assume we have 3 nodes named: Eager (does not support lazy parsing), Bug (contains the bug), Fixed (contains the bug fix). 

Bug -> Fixed -> Eager would fail.

However, any interactions between them

Eager <-> Fixed (Eager contains the parsed object and send it over the wire)
Bug <-> Fixed (Bug supports lazy parsing, so no need to parse on serialization)
Eager <-> Bug (Bug contains the code that can parse right before serialization)

Reply:
I might have a "simplified" view. Still I think that it would get simpler if you use the parsed query in this place. Therefore queryProvider always has both the parsed and surface query and in the case of a broken query only the surface form, in this case the datafeed is broken anyway. All the serialization and BWC quirks would need to be fixed in the serialization code in a way that no matter what queryProvider gets initialized correctly. The same applies for aggregations.

Reply:
@hendrikmuhs 
How could you parse a query from `Bug` version that only sends a `Map<String, Object>` over the wire? The serialization methods have no access to the `NamedXContentRegistry`

Reply:
I see that I misunderstood your previous answer. `QueryBuilder.parsedQuery` can be `null` in 1 case (version 6.6.0 - 6.6.2 if I read the labels correctly).

I suggest to define the API the way we want to have it for the future - ignoring the issue - 

```
public QueryBuilder getParsedQuery()
```

and have a fixup method which you call in the according places before calling `getParsedQuery()`

```
public fixupQueryBuilder(NamedXContentRegistry registry) {
  if queryProvider.getParsedQuery() == null {
  // this is the fixup code required because bla bla bla
  }
}
```

alternatively you cold have a bool `requiresParsingFix` in the QueryProvider class

The benefit of this approach: Going forward we want to get rid of this legacy. This way it is only a matter of deleting some LOC, otherwise you need another big re-factoring, which get easily forgotten and we have to live with this forever.

Correct me if I am wrong: Assuming this works and you get this in for 7.0 and 6.7, we could already delete this workaround on master (but maybe we do not want to do this right now to avoid any merge trouble, although merges shall be quite clean).

Reply:
having an outside system call `fixupQueryBuilder` and then `getParsedQuery` assumes outside classes know about the internal state of the object. It makes way more sense for a uniform api in accessing the lazily parsed objects. I do not want another class to have to know when to call `fixupQueryBuilder` vs when not to. 

Correct, in master (once all is backported and figured out), some of these weird work arounds can be refactored again. But I wanted to be close to what I am going to be backporting all the way back to a 6.6.x and 6.7.0